### PR TITLE
Multi general api

### PR DESCRIPTION
**Pull request by MjnMixael**
This PR reorganizes the API slightly, putting them roughly in top->bottom order as you'd progress through the UIs. It also moves Netgame, NetPlayer, Status Text, and Info Text up into the newer MultiGeneral category since they apply to more places than just MultiHostSetup or MultiJoinGame. While doing so it implements some additional validation for if those are accessed outside of proper Multi game states.

Finally this PR also implements `MultiGeneral.getChat()` and `.sendChat()`.

Be aware that for reviewing purposes, ui.cpp looks like a large diff but is actually nearly unchanged with the exception of the added code for MultiGeneral's getChat(), sendChat(), quitGame() and setPlayerState(). The rest of the changes that github is showing is just re-organized but unchanged code.

### CODE REVIEW

**Comment by MjnMixael** [comment]
EDIT: I added a couple more general methods I need here. quitGame and setPlayerState

File: code/missionui/chatbox.cpp [hunk 1]
```diff
@@ -402,28 +402,49 @@ size_t chatbox_get_msg_target_length(char *msg)
...
+		send_game_chat_packet(Net_player, msg, target, nullptr, temp);
+	}
+
+	chatbox_add_line(msg, MY_NET_PLAYER_NUM);
+}
+
 // automatically split up any input text, send it, and leave the remainder 
 void chatbox_autosplit_line()
 {
 	char msg[150];
```
Comment (notimaginative) [inline code/missionui/chatbox.cpp:436]: Going ahead and changing this to `char msg[CHATBOX_MAX_LEN]` would be a nice thing to do while your here. Looks like there is only the two places left to change the magic number.

File: code/missionui/chatbox.cpp [hunk 2]
```diff
@@ -433,22 +454,16 @@ void chatbox_autosplit_line()
...
+		// check if this message is supposed to have a recipient
+		int target = chatbox_get_msg_target_type(msg);
+		size_t target_length = chatbox_get_msg_target_length(msg);
 
 		if (target != MULTI_MSG_ALL) {
 			// we need to add the target the message is going to before we add the rest of the string
+			char temp[150];
 			strncpy(temp, msg, target_length);
  			temp[target_length] = ' '; 
  			temp[target_length+1] = '\0';
```
Comment (notimaginative) [inline code/missionui/chatbox.cpp:469]: As mentioned in the earlier change request, the size of `temp` should be de-magic-numbered here, as well as moving `target_length`.

File: code/missionui/chatbox.h [hunk 3]
```diff
@@ -59,6 +59,9 @@ int chatbox_scroll_down();
 // clear the contents of the chatbox
 void chatbox_clear();
 
+// send a message to the chatbox
+void chatbox_send_msg(char msg[150]);
```
Comment (notimaginative) [inline code/missionui/chatbox.h:63]: As mentioned earlier, this should be a ptr to `msg`.

File: code/scripting/api/libs/ui.cpp [hunk 4]
```diff
@@ -3200,135 +3283,255 @@ ADE_FUNC(setGameType,
...
-ADE_LIB_DERIV(l_Active_Games, "ActiveGames", nullptr, nullptr, l_UserInterface_MultiJoinGame);
-ADE_INDEXER(l_Active_Games,
-	"number Index",
-	"Array of active games",
-	"active_game",
-	"active game handle, or invalid handle if index is invalid")
+ADE_FUNC(runNetwork,
+	l_UserInterface_MultiSync,
+	nullptr,
+	"Runs the network required commands to update the lists and run the chat",
```
Comment (notimaginative) [inline code/scripting/api/libs/ui.cpp:3502]: This only covers multi sync, correct? So not list updates and chat.

**Review by notimaginative — CHANGES_REQUESTED** [review]

File: code/scripting/api/objs/multi_objects.cpp [hunk 5]
```diff
@@ -113,6 +118,15 @@ netgame_info* net_game_h::getNetgame() const
 	return &Netgame;
 }
 
+bool net_game_h::isValid() const
+{
+	if (!(Game_mode & GM_MULTIPLAYER) && (Netgame.name[0] != '\0')) {
```
[the remark this answers is not in the document]
Comment (MjnMixael) [reply code/scripting/api/objs/multi_objects.cpp]: I could have sworn I tested that.. but you're right. That is an error.

File: code/missionui/chatbox.h (hunk 3 above)
Comment (MjnMixael) [reply code/missionui/chatbox.h]: char* and CHATBOX_MAX_LEN all handled, I think.

File: code/scripting/api/libs/ui.cpp (hunk 4 above)
Comment (MjnMixael) [reply code/scripting/api/libs/ui.cpp]: Yup. Good catch.

File: code/missionui/chatbox.cpp [hunk 6]
```diff
@@ -402,28 +402,49 @@ size_t chatbox_get_msg_target_length(char *msg)
...
+	// if I'm the server, then broadcast the packet
+	chatbox_recall_add(msg);
+	if (target != MULTI_MSG_EXPR) {
+		send_game_chat_packet(Net_player, msg, target);
+	} else {
+		// copy the name of the player the message is being sent to
+		char temp[150];
+		strncpy(temp, msg + 1, target_length - 2);
+		temp[target_length - 2] = '\0';
+		send_game_chat_packet(Net_player, msg, target, nullptr, temp);
```
[the remark this answers is not in the document]
Comment (MjnMixael) [reply code/missionui/chatbox.cpp]: OK, target_length has been moved, too.

**Review by notimaginative — APPROVED** [review]

**Pull request merged**